server/pulp/server/db/querysets.py

```diff
@@ -1,10 +1,44 @@
-from mongoengine.queryset import DoesNotExist, QuerySet
+from mongoengine.queryset import DoesNotExist, QuerySetNoCache
 from pymongo import ASCENDING
 
 from pulp.server import exceptions as pulp_exceptions
 
 
-class CriteriaQuerySet(QuerySet):
+class CustomQuerySet(QuerySetNoCache):
+    """
+    All custom QuerySet classes should inherit from this class rather than QuerySet
+    or QuerySetNoCache. There are two reasons for this. Firstly, all QuerySets should
+    not cache their query results by default, so QuerySetNoCache must be inherited.
+    Secondly, QuerySetNoCache's implementation of the `cache` instance method
+    explicitly returns an instance of QuerySet which effectively removes any
+    customized behavior for that QuerySet.
+    """
+
+    def cache(self):
+        """
+        Currently, there is no reason to have a caching version of any of our
+        custom QuerySet classes. If that ever changes, this method should be
+        implemented to return a caching version of the custom QuerySet. This
+        implementation exists in order to avoid bugs resulting from the
+        unexpected default behavior of `cache`.
+
+        As of mongoengine 0.10, the default implementation of ``cache`` cannot
+        be used as it explicitly creates a QuerySet object, thus removing any
+        custom behavior of the custom QuerySet.
+        """
+        raise NotImplementedError()
+
+    def __len__(self):
```

Comment:
Jeremy Cline wrote:
> One thing to note is that although the caching version of QuerySet
> provides an implementation of |__len__|, the non-caching variety does
> not. Personally I feel like |__len__| is the "obvious way" to get the
> length of an iterable, but given that upstream went with both |count|
> and occasionally |__len__| (with different behaviors for each)...

Ah that does sway my opinion. I'd prefer not to implement __len__ here
if the superclass does not implement it.

-- 
Randy Barlow
irc:   bowlofeggs
phone: +1-919-890-8150